File: sdk/src/main/java/com/meniga/sdk/converters/MenigaTransactionsConverter.java
package com.meniga.sdk.converters;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import com.meniga.sdk.helpers.GsonProvider;
import com.meniga.sdk.models.transactions.MenigaComment;
import com.meniga.sdk.models.transactions.MenigaTransaction;
import com.meniga.sdk.models.transactions.MenigaTransactionPage;

import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.List;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class MenigaTransactionsConverter extends MenigaConverter {

	@Override
	public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, final Retrofit retrofit) {
		Type typeOfListTransactions = new TypeToken<List<MenigaTransaction>>() {
		}.getType();
		Type typeOfMenigaTransactionPage = new TypeToken<MenigaTransactionPage>() {
		}.getType();
		Type typeOfTransaction = new TypeToken<MenigaTransaction>() {
		}.getType();

		if (typeOfListTransactions.equals(type)) {
			return new Converter<ResponseBody, Object>() {
				@Override
				public Object convert(ResponseBody resBody) throws IOException {
					Gson gson = GsonProvider.getGsonBuilder();

					List<MenigaTransaction> menigaTransactions = gson.fromJson(getAsArray(resBody.byteStream()), new TypeToken<List<MenigaTransaction>>() {
					}.getType());
					for (MenigaTransaction trans : menigaTransactions) {
						if (trans.getComments() == null) {
							continue;
						}
						for (MenigaComment comm : trans.getComments()) {
							comm.setTransactionId(trans.getId());
						}
					}

					return menigaTransactions;
				}
			};
		} else if (typeOfTransaction.equals(type)) {
			return new Converter<ResponseBody, Object>() {
				@Override
				public Object convert(ResponseBody resBody) throws IOException {
					Gson gson = GsonProvider.getGsonBuilder();

					MenigaTransaction menigaTransaction = gson.fromJson(getAsObject(resBody.byteStream()), MenigaTransaction.class);
					if (menigaTransaction.getComments() != null) {
						for (MenigaComment comm : menigaTransaction.getComments()) {
							comm.setTransactionId(menigaTransaction.getId());
						}
					}

					return menigaTransaction;
				}
			};
		} else if (typeOfMenigaTransactionPage.equals(type)) {
			return new Converter<ResponseBody, Object>() {
				@Override
				public Object convert(ResponseBody resBody) throws IOException {
					Gson gson = GsonProvider.getGsonBuilder();

					InputStreamReader isr = new InputStreamReader(resBody.byteStream());
					JsonElement element;
					try {
						element = new JsonParser().parse(isr);
					} finally {
						isr.close();
					}

					List<MenigaTransaction> menigaTransactions = gson.fromJson(getAsArray(element), new TypeToken<List<MenigaTransaction>>() {
					}.getType());
					MenigaTransactionPage list = new MenigaTransactionPage();

					list.addAll(menigaTransactions);
					JsonObject object = element.getAsJsonObject();
					JsonObject meta = object.getAsJsonObject("meta");
					list.setTotalNumTransactions(meta.get("totalCount").getAsInt());

					return list;
				}
			};
		}

		return null;
	}

	@Override
	public Converter<?, RequestBody> requestBodyConverter(Type type, Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
		return null;
	}
}

File: sdk/src/main/java/com/meniga/sdk/converters/MenigaTransactionsConverter.kt
/*
 * Copyright 2017-2018 Meniga Iceland Inc.
 */
package com.meniga.sdk.converters

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.meniga.sdk.helpers.GsonProvider
import com.meniga.sdk.helpers.get
import com.meniga.sdk.helpers.type
import com.meniga.sdk.models.accounts.MenigaAccount
import com.meniga.sdk.models.merchants.MenigaMerchant
import com.meniga.sdk.models.transactions.MenigaTransaction
import com.meniga.sdk.models.transactions.MenigaTransactionPage
import com.meniga.sdk.models.transactions.setAccount
import com.meniga.sdk.models.transactions.setMerchant
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

class MenigaTransactionsConverter : MenigaConverter() {
    override fun responseBodyConverter(type: Type, annotations: Array<Annotation>, retrofit: Retrofit): Converter<ResponseBody, *>? {
        val gson = GsonProvider.getGsonBuilder()
        when (type) {
            type<List<MenigaTransaction>>() -> return Converter<ResponseBody, Any> { responseBody ->
                val (data, _, included) = MenigaConverter.getAsArrayApiResponse(responseBody.byteStream())
                val (transactions, accounts, merchants) = gson.getAccountsMerchantsAndTransactions(data, included)
                transactions
                        .onEach {
                            it.setAccount(accounts.find { account -> account.id == it.accountId })
                            it.setMerchant(merchants.find { merchant -> merchant.id == it.merchantId })
                            it.updateCommentsTransactionId()
                        }
            }
            type<MenigaTransaction>() -> return Converter<ResponseBody, Any> { responseBody ->
                val (data, _, included) = MenigaConverter.getAsObjectApiResponse(responseBody.byteStream())
                val transaction = gson.fromJson(data, MenigaTransaction::class.java)
                included?.let {
                    transaction.setAccount(gson.get(it, "account"))
                    transaction.setMerchant(gson.get(it, "merchant"))
                }
                transaction.apply {
                    updateCommentsTransactionId()
                }
            }

            type<MenigaTransactionPage>() -> return Converter<ResponseBody, Any> { responseBody ->
                val (data, meta, included) = MenigaConverter.getAsArrayApiResponse(responseBody.byteStream())
                val transactionPage = MenigaTransactionPage()
                val (transactions, accounts, merchants) = gson.getAccountsMerchantsAndTransactions(data, included)
                transactions.forEach { transaction ->
                    transaction.setAccount(accounts.find { it.id == transaction.accountId })
                    transaction.setMerchant(merchants.find { it.id == transaction.merchantId })
                    transaction.updateCommentsTransactionId()
                }
                transactionPage.apply {
                    addAll(transactions)
                    totalNumTransactions = meta?.get("totalCount")?.asInt ?: 0
                }
            }
            else -> return null
        }
    }

    private fun Gson.getAccountsMerchantsAndTransactions(data: JsonArray, included: JsonObject?)
            : Triple<List<MenigaTransaction>, List<MenigaAccount>, List<MenigaMerchant>> =
            Triple(get<List<MenigaTransaction>>(data).orEmpty(),
                    included?.let { get<List<MenigaAccount?>>(it, "accounts") }
                            .orEmpty()
                            .filterNotNull(),
                    included?.let { get<List<MenigaMerchant?>>(it, "merchants") }
                            .orEmpty()
                            .filterNotNull())

    private fun MenigaTransaction.updateCommentsTransactionId() {
        comments?.forEach { it.transactionId = id }
    }
}
